url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jutils/signalling/SignallingUtil.java

```diff
@@ -0,0 +1,916 @@
+     * @return The telephone-subscriber part of a tel: URI
+     */
+    public static String convertStringToTelephoneSubscriber(final String digits,
+            final TelnoType typePlan, final String phoneContextParameter,
+            final char phoneContextSeparator, final String defaultPhoneContext,
+            final boolean permitHexDigits) {
+        StringBuilder sb = new StringBuilder();
+
+        if (typePlan == TelnoType.INTERNATIONAL
+                && (permitHexDigits || DIGITS.matcher(digits).matches())) {
+            // global-number
+            sb.append('+').append(digits);
+            return sb.toString();
+        } else {
+            // local-number
+            String phoneContext = "";
+            int phoneContextIndex = digits.indexOf(phoneContextSeparator);
+            if (phoneContextIndex != -1) {
+                // The telno has an embedded phone-context
+                phoneContext = digits.substring(0, phoneContextIndex);
+                sb.append(digits.substring(phoneContextIndex + 1));
+                if (phoneContextIndex > 0
+                        && (permitHexDigits || DIGITS.matcher(phoneContext).matches())) {
+                    // The embedded phone-context is a valid global-number
+                    sb.append(';').append(phoneContextParameter).append('=');
+                    sb.append('+').append(phoneContext);
+                    return sb.toString();
+                }
+            } else {
+                // The telno does not have an embedded phone-context
+                sb.append(digits);
+            }
+
+            // Missing or invalid embedded phone-context
+            if (defaultPhoneContext == null) {
+                // No default phone-context so can't generate a valid URI
+                return null;
+            } else if (!defaultPhoneContext.isEmpty()) {
+                // Use the default phone-context
+                sb.append(';').append(phoneContextParameter).append('=');
+                sb.append(defaultPhoneContext);
+            }
+            // Strictly speaking a local-number without a phone-context is
+            // invalid, but we allow it if we've been explicitly given an empty
+            // default phone-context
+            return sb.toString();
+        }
+    }
+
+    /**
+     * Converts a generic telno into a String of digits. The hex digits A-F will
+     * be converted according to the <code>permitStarHash</code> setting.
+     *
+     * @param telno
+     *            The telno to convert into a string of digits
+     * @param permitStarHash
+     *            True if the telno digits 'ABCDEF' should be converted to the
+     *            characters '*#ABCD'; false if the digits should be converted
+     *            to the characters 'ABCDEF'
+     * @return A String containing the digits in the telno
+     */
+    public static String convertGenericTelnoToString(final GenericTelno telno,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 45 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.